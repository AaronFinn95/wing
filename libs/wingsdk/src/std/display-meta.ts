/**
 * Display meta component.
 */
export class DisplayMetaComponent {
  constructor(
    /**
     * The component type.
     */
    public readonly type: string,
    /**
     * The component props.
     */
    public readonly props: { [key: string]: any }
  ) {}
}

/**
 * Display meta link props
 */
export interface LinkProps {
  /**
   * The link href.
   */
  readonly href: string;
  /**
   * The link text.
   */
  readonly text: string;
}

/**
 * Display meta link component.
 */
export class DisplayMetaLink extends DisplayMetaComponent {
  constructor(props: LinkProps) {
    super("link", props);
  }
}

/**
 * Display meta
 */
export class DisplayMeta {
  private readonly _components: DisplayMetaComponent[];
  constructor() {
    this._components = [];
  }

  /**
   * Returns the meta components.
   */
  public get components(): DisplayMetaComponent[] | undefined {
    return [...this._components];
  }

  private addComponent(component: DisplayMetaComponent) {
    this._components.push(component);
  }

  /**
   * Add a link to the array of meta components.
   * @param href
   * @param text
   */
  public addLink(href: string, text: string) {
    this.addComponent(new DisplayMetaLink({ href, text }));
  }
}
