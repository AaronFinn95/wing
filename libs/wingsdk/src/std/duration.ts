/**
 * Represents a length of time.
 */
export class Duration {
  /**
   * Create a Duration representing an amount of minutes
   *
   * @param amount the amount of Minutes the `Duration` will represent.
   * @returns a new `Duration` representing `amount` Minutes.
   */
  public static fromMinutes(amount: number): Duration {
    return new Duration(amount * 60 * 1000);
  }

  /**
   * Create a Duration representing an amount of hours
   *
   * @param amount the amount of Hours the `Duration` will represent.
   * @returns a new `Duration` representing `amount` Hours.
   */
  public static fromHours(amount: number): Duration {
    return new Duration(amount * 60 * 60 * 1000);
  }

  /**
   * Create a Duration representing an amount of seconds
   *
   * @param amount the amount of Seconds the `Duration` will represent.
   * @returns a new `Duration` representing `amount` Seconds.
   */
  public static fromSeconds(amount: number): Duration {
    return new Duration(amount * 1000);
  }

  /**
   * Create a Duration representing an amount of milliseconds
   *
   * @param amount the amount of milliseconds the `Duration` will represent.
   * @returns a new `Duration` representing `amount` milliseconds.
   */
  public static fromMilliseconds(amount: number): Duration {
    return new Duration(amount);
  }

  /**
   * Return the total number of milliseconds in this Duration
   *
   * @returns the value of this `Duration` expressed in Milliseconds.
   */
  public readonly milliseconds: number;

  private constructor(milliseconds: number) {
    this.milliseconds = milliseconds;
  }

  /**
   * Return the total number of minutes in this Duration
   *
   * @returns the value of this `Duration` expressed in Minutes.
   */
  public get minutes() {
    return this.seconds / 60;
  }

  /**
   * Return the total number of hours in this Duration
   *
   * @returns the value of this `Duration` expressed in Hours.
   */
  public get hours() {
    return this.minutes / 60;
  }

  /**
   * Return the total number of milliseconds in this Duration
   *
   * @returns the value of this `Duration` expressed in Milliseconds.
   */
  public get seconds() {
    return this.milliseconds / 1000;
  }
}
