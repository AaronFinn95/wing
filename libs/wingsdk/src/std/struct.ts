import { T1 } from "./generics";
import { Code, InflightClient } from "../core";
import { Json } from "./json";

/**
 * Base class for all structs
 * 
 * @typeparam T1
 */
export class BaseStruct {
  /**
   * @internal
   */
  public static _toInflightType(): Code {
    return InflightClient.forType(__filename, this.name);
  }

  private constructor() {}

  /**
   * Converts a json to struct
   * 
   * @macro ((j, s) => {for (const key in s) {const expectedType = s[key];const actualType = typeof j[key]; if (!expectedType.includes(actualType)) {throw new Error(`key "${key}" expected type of "${expectedType}", but received type of: "${actualType}"`);}}; return j;})($args$, $self$)
   */
  public static fromJson(json: Json): T1 {
    json;
    throw new Error("Macro")
  }
}