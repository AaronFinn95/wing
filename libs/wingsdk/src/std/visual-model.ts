/**
 * Visual Model Component Type.
 */
export type ComponentType = "link";

/**
 * Visual Model Component.
 */
export interface VisualModelComponent {
  /**
   * The component type.
   */
  readonly type: ComponentType;

  /**
   * The component props.
   */
  readonly props: { [key: string]: any };
}

/**
 * Link Component props
 */
export interface LinkComponent {
  /**
   * The link href.
   */
  readonly href: string;
  /**
   * The link text.
   */
  readonly text: string;
}

/**
 * Visual Model. Used to generate the visual model for a resource.
 */
export interface VisualModel {
  /**
   * Visual Model Components.
   */
  readonly components: VisualModelComponent[];
}

/**
 * Visual Model Builder. Used to build a visual model.
 */
export class VisualModelBuilder {
  private readonly components: VisualModelComponent[] = [];

  /**
   * Adds a link component to the visual model.
   * @param href
   * @param text
   */
  public addLink(href: string, text: string): void {
    this.components.push({
      type: "link",
      props: {
        href,
        text,
      },
    });
  }

  /**
   * Builds the visual model.
   * @returns the visual model.
   */
  public buildVisualModel(): VisualModel {
    return {
      components: this.components,
    };
  }
}
