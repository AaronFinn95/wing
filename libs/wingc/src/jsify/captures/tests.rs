use super::ClassCaptures;
use crate::{
	ast::{Class, Scope},
	diagnostic::{found_errors, get_diagnostics},
	jsify::JSifier,
	partial_compile,
	visit::{self, Visit},
};
use insta::assert_snapshot;
use std::{
	env,
	fs::{self, File},
	io::Write,
};

// -----------------------------------------------------------------------------

#[test]
fn read_primitive_value() {
	assert_snapshot!(capture_ok(
		r#"
    let x = "my_string";

    test "test" {
      log(x);
    }
    "#,
	));
}

#[test]
fn access_methods_and_properties_on_collections() {
	assert_snapshot!(capture_ok(
		r#"
    let x = [1,2,3];
    let y = ["hello"];

    test "test" {
      assert(x.length == 3);
      assert(y.at(0) == "hello");
    }
    "#
	));
}

#[test]
fn access_property_on_primitive() {
	assert_snapshot!(capture_ok(
		r#"
    let s = "hello";

    test "test" {
      assert(s.length == 5);
    }
    "#
	));
}

#[test]
fn access_property_on_value_returned_from_collection() {
	assert_snapshot!(capture_ok(
		r#"
    let s = { "hello": "123" };

    test "test" {
      assert(s.get("hello").length == 3);
    }
    "#
	));
}

#[test]
fn calls_methods_on_preflight_object() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;
    let b = new cloud.Bucket();

    test "test" {
      b.put("hello", "world");
      b.list();
    }
    "#
	));
}

#[test]
fn fails_on_reassignable_variable() {
	assert_snapshot!(capture_fail(
		r#"
    let var x = "my_string";

    test "test" {
      log(x);
    }
    "#,
	));
}

#[test]
fn fails_on_mutable_collections() {
	assert_snapshot!(capture_fail(
		r#"
    let x = MutArray<str>["hello", "world"];

    test "test" {
      assert(x.at(0) == "hello");
    }
    "#,
	));
}

#[test]
fn enum_value() {
	assert_snapshot!(capture_ok(
		r#"
    enum MyEnum { B, C }

    test "test" {
      assert(MyEnum.B != MyEnum.C);
    }
    "#,
	));
}

#[test]
fn static_on_std_type() {
	assert_snapshot!(capture_ok(
		r#"
    test "test" {
      log(str.fromJson("hello"));
      assert(Json.values(Json {}).length == 0);
    }
    "#,
	));
}

#[test]
fn preflight_value_field() {
	assert_snapshot!(capture_ok(
		r#"
    class MyType {
      name: str;
      last: str;

      init() {
        this.name = "hello";
        this.last = "world";
      }
    }

    let t = new MyType();

    test "test" {
      log(t.name);
      assert(t.name.length > 0);
      log(t.last);
    }
    "#,
	));
}

#[test]
fn inflight_field_from_inflight() {
	assert_snapshot!(capture_ok(
		r#"
    class MyType {
      inflight s: str;
      inflight init() {
        this.s = "hey there!";
      }
      inflight foo() {
        log(this.s);
      }
    }
    "#,
	));
}

#[test]
fn var_inflight_field_from_inflight() {
	assert_snapshot!(capture_ok(
		r#"
    class MyType {
      inflight var s: str;
      inflight init() {
        this.s = "hey there!";
      }
      inflight foo() {
        log(this.s);
        this.s = "hello!";
      }
    }
    "#,
	));
}

#[test]
fn preflight_object_through_property() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;

    class MyType {
      b: cloud.Bucket;
      init() {
        this.b = new cloud.Bucket();
      }
    }

    let t = new MyType();

    test "test" {
      t.b.put("hello", "world");
    }
    "#,
	));
}

#[test]
fn fails_on_preflight_static() {
	assert_snapshot!(capture_fail(
		r#"
    class MyType {
      static staticMethod(): str {}
    }

    test "test" {
      log(MyType.staticMethod());
    }
    "#,
	));
}

#[test]
fn reference_static_inflight() {
	assert_snapshot!(capture_ok(
		r#"
    class MyType {
      static inflight staticMethod(): str {}
    }

    test "test" {
      log(MyType.staticMethod());
    }
    "#,
	));
}

#[test]
fn reference_from_static_inflight() {
	assert_snapshot!(capture_ok(
		r#"
    let s = "hello";

    class MyType {
      static inflight staticMethod(): str {
        return s;
      }
    }
    "#,
	));
}

#[test]
fn reference_preflight_object_from_static_inflight() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;

    let q = new cloud.Queue();

    class MyType {
      static inflight addToQueue(m: str) {
        q.push(m);
      }
    }
    "#,
	));
}

#[test]
fn new_inflight_object() {
	assert_snapshot!(capture_ok(
		r#"
    inflight class Foo {}

    test "test" {
      new Foo();
    }
    "#,
	));
}

#[test]
fn reference_preflight_fields() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;
    
		class MyType {
      s: str;
      b: cloud.Bucket;

      init() { 
        this.s = "hello"; 
        this.b = new cloud.Bucket();
      }

      inflight boom() {
        assert(this.s.length > 0);
        assert(this.b.list().length > 0);
      }

      inflight bam() {
        this.b.put("hello", "world");
        this.b.get("aaa");
      }
    }
		"#,
	));
}

#[test]
fn transitive_reference() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;
    
		class MyType {
      b: cloud.Bucket;

      init() {
        this.b = new cloud.Bucket();
      }

      inflight isEmpty(): bool {
        return this.b.list().length == 0;
      }

      inflight checkIfEmpty() {
        if this.isEmpty() {
          log("empty!");
        }
      }
    }

    let t = new MyType();
    test "test" {
      t.checkIfEmpty();
    }
		"#,
	));
}

#[test]
fn ref_std_macro() {
	assert_snapshot!(capture_ok(
		r#"
    let arr = [1,2,3];

    test "test" {
      assert(arr.length == 3);
    }
		"#,
	));
}

#[test]
fn fails_when_referencing_this_from_static() {
	assert_snapshot!(capture_fail(
		r#"
		class MyType {
      static inflight hello() {
        this.print();
      }
    }
		"#,
	));
}

#[test]
fn transitive_reference_via_static() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;

    let b = new cloud.Bucket();
    
		class MyType {
      static inflight putInBucket() {
        b.put("in", "bucket");
      }
    }

    class YourType {
      inflight putIndirect() {
        MyType.putInBucket();
      }
    }

    let t = new YourType();
    test "test" {
      t.putIndirect();
    }
		"#,
	));
}

#[test]
fn fails_if_referencing_unknown_field() {
	assert_snapshot!(capture_fail(
		r#"
		class MyInflightClass {
      inflight putInBucket() {
        this.b.put("in", "bucket");
      }
    }
		"#,
	));
}

#[test]
fn transitive_reference_via_inflight_class() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;

    let b = new cloud.Bucket();

		inflight class MyInflightClass {
      putInBucket() {
        b.put("in", "bucket");
      }
    }

    test "test" {
      let obj = new MyInflightClass();
      obj.putInBucket();
    }
		"#,
	));
}

#[test]
fn reference_inflight_class() {
	assert_snapshot!(capture_ok(
		r#"
    inflight class Foo {
      static a(): str { return "a"; }
    }

    test "test" {
      log(Foo.a());
    }
    "#
	));
}

#[test]
fn inline_inflight_class() {
	assert_snapshot!(capture_ok(
		r#"
    bring cloud;
    class Foo {
      b: cloud.Bucket;
      q: cloud.Queue;

      init() {
        this.b = new cloud.Bucket();
        this.q = new cloud.Queue();

        this.q.setConsumer(inflight () => {
          this.b.put("in", "bucket");
        });
      }
    }
    "#
	));
}

#[test]
fn temp() {
	assert_snapshot!(capture_ok(
		r#"
    let jsonObj1 = Json { key1: "value1" };
 
    test "test" {
      log(Json.stringify(jsonObj1));
    }  
    "#
	));
}

// -----------------------------------------------------------------------------
// test utility functions

fn capture_ok(code: &str) -> String {
	let snap = capture_report(code);
	assert!(!found_errors());
	snap
}

fn capture_fail(code: &str) -> String {
	let snap = capture_report(code);
	assert!(found_errors());
	snap
}

/// Compiles `code` and returns the capture scanner results as a string that can be snapshotted
fn capture_report(code: &str) -> String {
	let workdir = tempfile::tempdir().unwrap();
	let path = workdir.path().join("main.w");

	// NOTE: this is needed for debugging to work regardless of where you run the test
	env::set_current_dir(env!("CARGO_MANIFEST_DIR")).unwrap();

	// convert tabs to 2 spaces
	let code = code.replace("\t", "  ");

	let mut f = File::create(&path).unwrap();
	f.write_all(code.as_bytes()).unwrap();

	let source_path = path.as_path();
	let (scope, types) = partial_compile(source_path);

	let mut snap = vec![];

	if !found_errors() {
		let classes = find_classes(&scope);
		for class in classes {
			snap.push("======================================================".into());
			snap.push(format!("Captured by {}:", class.name.name));
			let captures = ClassCaptures::scan(&types, &class);
			snap.push(captures.to_string());
		}
	}

	if found_errors() {
		snap.push("======================================================".into());
		snap.push("Errors:".into());
		for diag in get_diagnostics() {
			let span = diag
				.span
				.map(|f| format!("{}:{}", f.start.line, f.start.col))
				.unwrap_or_default();
			snap.push(format!(" - at {} | {}", span, diag.message));
		}
	}

	let mut jsify = JSifier::new(&types, "default", &workdir.path(), false);
	jsify.jsify(&scope);
	jsify.emit_files(workdir.path());

	snap.push("======================================================".into());
	snap.push("Files:".into());
	snap.push("".into());

	for entry in fs::read_dir(workdir.path()).unwrap() {
		let Ok(entry) = entry else {
      continue;
    };

		let contents = fs::read_to_string(entry.path()).unwrap();
		let x = entry.file_name().as_os_str().to_string_lossy().to_string();

		snap.push(x);
		snap.push("------------------------------------------------------".into());
		snap.push(contents);
	}

	return snap.join("\n");
}

fn find_classes(scope: &Scope) -> Vec<&Class> {
	// find the "myMethod" scope
	struct FindClassVisitor<'a> {
		results: Vec<&'a Class>,
	}

	impl<'a> Visit<'a> for FindClassVisitor<'a> {
		fn visit_class(&mut self, node: &'a Class) {
			self.results.push(&node);
			visit::visit_class(self, node);
		}
	}

	let mut visitor = FindClassVisitor { results: vec![] };

	visitor.visit_scope(scope);
	return visitor.results;
}
