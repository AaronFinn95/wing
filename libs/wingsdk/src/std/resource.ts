import { Construct, IConstruct } from "constructs";
import { Datetime } from "./datetime";
import { Duration } from "./duration";
import { App, Bindings, Connections } from "../core";

/**
 * A resource that can run inflight code.
 * @skipDocs
 */
export interface IInflightHost extends IConstruct {}

/**
 * A resource with inflight operations.
 */
export interface IInflightConstruct extends IConstruct {
  /**
   * Return a code snippet that can be used to reference this resource inflight.
   *
   * Note this code snippet may by async code, so it's unsafe to run it in a
   * constructor or other sync context.
   *
   * @internal
   */
  _toInflight(): string;
}

/**
 * This interface contains documentation for methods that the Wing compiler generates but may
 * not be needed by all constructs.
 */
// @ts-ignore allow unused declaration
interface IExtraInflightOperations {
  /**
   * Binds the resource to the host so that it can be used by inflight code.
   *
   * If a construct implements this method, it will not be called no more than
   * once with a given `host`.
   *
   * If a construct doesn't implement this method, it just means that none of its
   * inflight operations add any requirements on the host (like IAM permissions or
   * environment variables).
   *
   * This method shouldn't call `bind` on any other constructs besides its parent `super`.
   */
  bind(host: IInflightHost, ops: string[]): void;

  /**
   * Return a list of supported inflight operations supported by the resource.
   *
   * If a construct doesn't implement this method, it's assumed that the construct
   * doesn't support any inflight operations.
   */
  _getInflightOps(): string[];

  /**
   * Register any relationships between
   * the class and any transitively referenced classes. For example, if calling
   * the inflight method "foo" on a class requires calling the inflight methods
   * "bar" and "baz" on another class, then the compiler might generate:
   *
   * ```
   * _registerBind(host, ops) {
   *   if (ops.includes("foo")) {
   *     this._bar._registerBind(host, ["bar", "baz"]);
   *   }
   * }
   * ```
   *
   * The method's implementation shouldn't contain any actual binding logic
   * (like adding IAM permissions or environment variables to the host).
   *
   * If the method isn't implemented, it can be assumed that the class doesn't
   * reference any other classes.
   */
  _registerBind(host: IInflightHost, ops: string[]): void;

  /**
   * (static method) Register that the resource type needs to be bound to the host for the given
   * operations. A type being bound to a host means that that type's static members
   * will be bound to the host.
   */
  _registerTypeBind(host: IInflightHost, ops: string[]): void;
}

/**
 * Shared behavior between all Wing SDK resources.
 * @skipDocs
 */
export class Resource {
  /**
   * Register a binding between an object (either data or class or instance) and a host.
   * All of the registered bindings will be aggregated, so that when synthesis is
   * performed, all of the necessary `bind()` calls will be invoked.
   *
   * - Primitives, Duration, and Datetime objects are ignored.
   * - Arrays, sets and maps and structs (Objects) are recursively bound.
   * - Class instances are bound to the host by calling their bind() method.
   * - (TODO) Classes and their static methods are bound to their host by calling their bindType() method.
   *
   * @param obj The object to bind.
   * @param host The host to bind to
   * @param ops The set of operations that may access the object
   *
   * @internal
   */
  public static _registerBindObject(
    obj: any,
    host: IInflightHost,
    ops: string[] = []
  ): void {
    const tokens = App.of(host)._tokens;
    if (tokens.isToken(obj)) {
      return tokens.bindValue(host, obj);
    }

    switch (typeof obj) {
      case "string":
      case "boolean":
      case "number":
      case "undefined":
        return;

      case "object":
        if (Array.isArray(obj)) {
          obj.forEach((item) => Resource._registerBindObject(item, host));
          return;
        }

        if (obj instanceof Duration) {
          return;
        }

        if (obj instanceof Datetime) {
          return;
        }

        if (obj instanceof Set) {
          return Array.from(obj).forEach((item) =>
            Resource._registerBindObject(item, host)
          );
        }

        if (obj instanceof Map) {
          Array.from(obj.values()).forEach((item) =>
            Resource._registerBindObject(item, host)
          );
          return;
        }

        // if the object is a resource (i.e. has a "bind" method"), register a binding between it and the host.
        if (Construct.isConstruct(obj)) {
          // Explicitly register the resource's `$inflight_init` op, which is a special op that can be used to makes sure
          // the host can instantiate a client for this resource.
          let extendedOps = [...ops, "$inflight_init"];

          const bindings = Bindings.of(obj);

          // For each operation, check if the host supports it. If it does, register the binding.
          const supportedOps = [
            ...((obj as any)._getInflightOps?.() ?? []),
            "$inflight_init",
          ];

          for (const op of extendedOps) {
            if (!supportedOps.includes(op)) {
              throw new Error(
                `Resource ${obj.node.path} does not support inflight operation ${op} (requested by ${host.node.path})`
              );
            }

            if (bindings.has(host, op)) {
              // this resource is already bound to the host for this operation
              continue;
            }

            // add the operation to the set of operations for the host so that we can avoid
            // infinite recursion.
            bindings.add(host, op);

            if (typeof (obj as any)._registerBind === "function") {
              (obj as any)._registerBind?.(host, [op]);
            }

            // add connection metadata
            Connections.of(obj).add({
              source: host,
              target: obj,
              name: op.endsWith("()") ? op : `${op}()`,
            });
          }
          return;
        }

        // structs are just plain objects
        if (obj.constructor.name === "Object") {
          Object.values(obj).forEach((item) =>
            Resource._registerBindObject(item, host, ops)
          );
          return;
        }
        break;

      case "function":
        // If the object is actually a resource type, call the type's _registerTypeBind static method
        if (typeof (obj as any)._registerTypeBind === "function") {
          obj._registerTypeBind(host, ops);
          return;
        }
        break;
    }

    throw new Error(
      `unable to serialize immutable data object of type ${obj.constructor?.name}`
    );
  }
}
