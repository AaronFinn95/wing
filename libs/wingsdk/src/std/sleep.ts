import { Duration } from "./duration";

export const sleep = async (duration: Duration): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, duration.milliseconds));
