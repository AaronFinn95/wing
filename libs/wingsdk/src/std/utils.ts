import { Duration } from "./duration";

export class Utils {
  public static async sleep(duration: Duration): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, duration.milliseconds));
  }
}
