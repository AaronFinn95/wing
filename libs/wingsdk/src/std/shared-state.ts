import { Construct } from "constructs";
import { IResource, Resource } from "./resource";
import { fqnForType } from "../constants";
import { App } from "../core";

/**
 * Global identifier for `SharedState`.
 */
export const SHARED_STATE_FQN = fqnForType("std.SharedState");

/**
 * A container for a value that can be shared and mutated across inflight consumers.
 * Only supported in the `sim` compilation target.
 *
 * @inflight `@winglang/sdk.std.ISharedStateClient`
 */
export abstract class SharedState extends Resource implements IResource {
  /**
   * Create a new SharedState instance.
   * @internal
   */
  public static _newSharedState(
    scope: Construct,
    id: string,
    initial: any
  ): SharedState {
    return App.of(scope).newAbstract(SHARED_STATE_FQN, scope, id, initial);
  }

  constructor(scope: Construct, id: string, initial: any) {
    super(scope, id);

    initial;

    this.display.title = "SharedState";
    this.display.description =
      "A container that can share a value across inflight consumers.";

    this._addInflightOps(SharedStateMethods.GET, SharedStateMethods.SET);
  }
}

/**
 * Inflight interface for `SharedState`.
 */
export interface ISharedStateClient {
  /**
   * Get the inner value.
   * @inflight
   */
  get(): Promise<any>;

  /**
   * Replace the inner value with a new value.
   * @inflight
   */
  set(value: any): Promise<void>;
}

/**
 * List of inflight operations available for `SharedState`.
 * @internal
 */
export enum SharedStateMethods {
  /** `SharedState.get` */
  GET = "get",
  /** `SharedState.set` */
  SET = "set",
}
