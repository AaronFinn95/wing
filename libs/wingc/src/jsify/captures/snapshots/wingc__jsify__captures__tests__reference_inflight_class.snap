---
source: libs/wingc/src/jsify/captures/tests.rs
expression: "capture_ok(r#\"\n    inflight class Foo {\n      static a(): str { return \"a\"; }\n    }\n\n    test \"test\" {\n      log(Foo.a());\n    }\n    \"#)"
---
======================================================
Captured by Foo:

======================================================
Captured by $Closure1:
Types:
  Foo = Foo

======================================================
Files:

inflight.$Closure1.js
------------------------------------------------------
module.exports = function({ Foo }) {
  class $Closure1 {
    constructor({  }) {
      const $obj = (...args) => this.handle(...args);
      Object.setPrototypeOf($obj, this);
      return $obj;
    }
    async handle()  {
      {console.log((await Foo.a()))};
    }
    async $inflight_init()  {
    }
  }
  return $Closure1;
}

main.w
------------------------------------------------------

    inflight class Foo {
      static a(): str { return "a"; }
    }

    test "test" {
      log(Foo.a());
    }
    
preflight.js
------------------------------------------------------
class Foo extends $stdlib.std.Resource {
  constructor(scope, id, ) {
    super(scope, id);
    this._addInflightOps("a", "constructor");
  }
  static _toInflightType(context) {
    return $stdlib.core.NodeJsCode.fromInline(`
      require("./inflight.Foo.js")({
      })
    `);
  }
  _toInflight() {
    return $stdlib.core.NodeJsCode.fromInline(`
      (await (async () => {
        const FooClient = ${Foo._toInflightType(this).text};
        const client = new FooClient({
        });
        if (client.$inflight_init) { await client.$inflight_init(); }
        return client;
      })())
    `);
  }
}
class $Closure1 extends $stdlib.std.Resource {
  constructor(scope, id, ) {
    super(scope, id);
    this._addInflightOps("handle", "$inflight_init");
    this.display.hidden = true;
  }
  static _toInflightType(context) {
    return $stdlib.core.NodeJsCode.fromInline(`
      require("./inflight.$Closure1.js")({
        Foo: Foo._toInflightType(context);
      })
    `);
  }
  _toInflight() {
    return $stdlib.core.NodeJsCode.fromInline(`
      (await (async () => {
        const $Closure1Client = ${$Closure1._toInflightType(this).text};
        const client = new $Closure1Client({
        });
        if (client.$inflight_init) { await client.$inflight_init(); }
        return client;
      })())
    `);
  }
}
this.node.root.new("@winglang/sdk.std.Test",std.Test,this,"test:test",new $Closure1(this,"$Closure1"));

inflight.Foo.js
------------------------------------------------------
module.exports = function({  }) {
  class Foo {
    static async a()  {
      return "a";
    }
     constructor()  {
    }
  }
  return Foo;
}

